plugins/parsers/bindata/parser.go

```diff
@@ -0,0 +1,205 @@
+			}
+		}
+	}
+
+	return &BinData{
+		MetricName:     metricName,
+		TimeFormat:     timeFormat,
+		Endiannes:      endiannes,
+		byteOrder:      byteOrder,
+		StringEncoding: stringEncoding,
+		Fields:         fields,
+		DefaultTags:    defaultTags,
+	}, nil
+}
+
+// SetDefaultTags implements Parser.SetDefaultTags()
+func (binData *BinData) SetDefaultTags(tags map[string]string) {
+	binData.DefaultTags = tags
+}
+
+// Parse implements Parser.Parse()
+func (binData *BinData) Parse(data []byte) ([]telegraf.Metric, error) {
+
+	fields := make(map[string]interface{})
+	reader := io.NewSectionReader(bytes.NewReader(data), 0, int64(len(data)))
+	var offset int64 = 0
+
+	for _, field := range binData.Fields {
+		if field.Type != "padding" {
+			fieldBuffer := make([]byte, field.Size)
+
+			if _, err := reader.ReadAt(fieldBuffer, offset); err != nil {
+				return nil, err
+			}
+
+			fieldType, ok := fieldTypes[strings.ToLower(field.Type)]
+			if !ok {
+				return nil, fmt.Errorf(`invalid field type %s`, field.Type)
+			}
+
+			switch fieldType.Name() {
+			case "string":
+				fields[field.Name] = string(fieldBuffer)
+			default:
+				fieldValue := reflect.New(fieldType)
+				byteReader := bytes.NewReader(fieldBuffer)
+				binary.Read(byteReader, binData.byteOrder, fieldValue.Interface())
+				fields[field.Name] = fieldValue.Elem().Interface()
+			}
+		}
+		offset += int64(field.Size)
+	}
+
+	metricTime, err := binData.getTime(fields)
+	if err != nil {
+		return nil, err
+	}
+
+	metric, err := metric.New(binData.MetricName, binData.DefaultTags,
+		fields, metricTime)
+	if err != nil {
+		return nil, err
+	}
+
+	return []telegraf.Metric{metric}, err
+}
+
+// ParseLine implements Parser.ParseLine()
+func (binData *BinData) ParseLine(line string) (telegraf.Metric, error) {
+	return nil, fmt.Errorf("BinData.ParseLine() not supported")
+}
+
+// Supported field types
+var fieldTypes = map[string]reflect.Type{
+	"bool":    reflect.TypeOf((*bool)(nil)).Elem(),
+	"uint8":   reflect.TypeOf((*uint8)(nil)).Elem(),
+	"int8":    reflect.TypeOf((*int8)(nil)).Elem(),
+	"uint16":  reflect.TypeOf((*uint16)(nil)).Elem(),
+	"int16":   reflect.TypeOf((*int16)(nil)).Elem(),
+	"uint32":  reflect.TypeOf((*uint32)(nil)).Elem(),
+	"int32":   reflect.TypeOf((*int32)(nil)).Elem(),
+	"uint64":  reflect.TypeOf((*uint64)(nil)).Elem(),
+	"int64":   reflect.TypeOf((*int64)(nil)).Elem(),
+	"float32": reflect.TypeOf((*float32)(nil)).Elem(),
+	"float64": reflect.TypeOf((*float64)(nil)).Elem(),
+	"string":  reflect.TypeOf((*string)(nil)).Elem(),
+	"padding": reflect.TypeOf((*[]byte)(nil)).Elem(),
+}
+
+func (binData *BinData) getTime(fields map[string]interface{}) (time.Time, error) {
+	nilTime := new(time.Time)
+	metricTime := time.Now()
+	timeValue := fields[timeKey]
+	if timeValue != nil {
+		var err error
+		switch binData.TimeFormat {
+		case "unix":
+			metricTime, err = internal.ParseTimestamp(binData.TimeFormat, int64(timeValue.(int32)), timezone)
```

Comment:
This will panic if the datatype of the `timeValue` field does not match what you expect here. The least line of defense would be IMO to check this in the `NewBinaryParser()` function by comparing the `type` of the `timeField` against the time-format.
But we here really want to check the type-assertion I think.

Reply:
Added a check in the NewBinDataParser